x/programs/cmd/simulator/src/lib.rs

```diff
@@ -227,105 +239,70 @@ impl Client {
     /// # Errors
     ///
     /// Returns an error if the serialization or plan fails.
-    pub fn run_plan(&self, plan: &Plan) -> Result<Vec<PlanResponse>, Box<dyn Error>> {
-        run_steps(self.path, plan)
-    }
-
-    /// Performs a single `Execute` step against the simulator and returns the result.
-    /// # Errors
-    ///
-    /// Returns an error if the serialization or single-[Step]-[Plan] fails.
-    pub fn execute_step(&self, key: &str, step: Step) -> Result<PlanResponse, Box<dyn Error>> {
-        let plan = &Plan {
-            caller_key: key.into(),
-            steps: vec![step],
-        };
-
-        run_step(self.path, plan)
+    pub fn run_plan(self, plan: Plan) -> Result<Vec<PlanResponse>, ClientError> {
+        let mut child = Command::new(self.path)
+            .arg("interpreter")
+            .arg("--cleanup")
+            .arg("--log-level")
+            .arg("error")
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()?;
+
+        let writer = child.stdin.take().ok_or("failed to open stdin").unwrap();
+        let reader = child.stdout.take().ok_or("failed to open stdout").unwrap();
+
+        let responses = BufReader::new(reader)
+            .lines()
+            .map(|line| -> Result<_, ClientError> {
+                let line = line.map_err(ClientError::Read)?;
+                let res = serde_json::from_str::<PlanResponse>(&line)
+                    .map_err(ClientError::Deserialization)?;
+                Ok(res)
+            });
```

Comment:
Use inference as much as possible. 
```suggestion
        let responses = BufReader::new(reader)
            .lines()
            .map(|line| serde_json::from_str(&line?).map_err(Into::into));
```

Reply:
This is looking much better!